### PR TITLE
Implement HTTP chunked Transfer-Encoding support

### PR DESCRIPTION
**Pull request by kacf**

### CODE REVIEW

File: src/common/http/platform/beast/http.cpp [hunk 1]
```diff
@@ -255,6 +255,37 @@ size_t GetContentLength(const PARSER &parser) {
...
+	const expected::ExpectedString &content_length,
+	const expected::ExpectedString &transfer_encoding) {
+	bool chunked;
+	if (transfer_encoding) {
+		if (transfer_encoding.value() != "chunked") {
+			return expected::unexpected(error::Error(
+				make_error_condition(errc::not_supported),
+				"Unsupported Transfer-Encoding: " + transfer_encoding.value()));
+		}
+		chunked = true;
```
Comment (oleorhagen) [inline src/common/http/platform/beast/http.cpp:268]: Q: Why not simply return here?

File: src/common/http_test.cpp [hunk 2]
```diff
@@ -672,6 +740,80 @@ TEST(HttpTest, TestResponseBody) {
...
+		},
+		[&received_body, &loop, &count](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+
+			vector<uint8_t> expected_body;
+			io::ByteWriter expected_writer(expected_body);
+			expected_writer.SetUnlimited(true);
+			io::Copy(expected_writer, *make_shared<BodyOfXes>());
+
+			ASSERT_EQ(received_body.size(), expected_body.size());
```
Comment (oleorhagen) [inline src/common/http_test.cpp:801]: Nit: `EXPECT`

File: src/common/http/platform/beast/http.cpp (hunk 1 above)
Comment (kacf) [reply src/common/http/platform/beast/http.cpp:268]: You're right, I guess my logic is a bit overkill. I'll do that.

File: src/common/http/platform/beast/http.cpp [hunk 3]
```diff
@@ -255,6 +255,31 @@ size_t GetContentLength(const PARSER &parser) {
...
+
+	if (content_length) {
+		auto length = common::StringToLongLong(content_length.value());
+		if (!length || length.value() < 0) {
+			return expected::unexpected(error::Error(
+				length.error().code,
+				"Content-Length contains invalid number: " + content_length.value()));
+		}
+		return length.value() > 0;
+	}
```
Comment (lluiscampos) [inline src/common/http/platform/beast/http.cpp:278]: We can use here also `GetContentLength`, right?

**Review by lluiscampos — APPROVED** [review]
It looks good!

As discussed privately, as long as this code does not add regressions on the current tests, we can merge this one and I'll investigate the `mender-gateway` tests failures independently.

File: src/common/http/platform/beast/http.cpp [hunk 4]
```diff
@@ -476,7 +486,7 @@ io::ExpectedAsyncReaderPtr Client::MakeBodyAsyncReader(IncomingResponsePtr resp)
 			"MakeBodyAsyncReader called while reading is in progress"));
 	}
 
-	if (response_body_length_ == 0) {
+	if (GetContentLength(*response_data_.http_response_parser_) == 0) {
```
[the remark this answers is not in the document]
Comment (kacf) [reply src/common/http/platform/beast/http.cpp]: Yes.... Actually, I'm wondering if that whole block is unnecessary. The `MakeBodyAsyncReader` call will return error if there is no body, so the decision can be entirely based on that.

File: src/common/http/platform/beast/http.cpp (hunk 3 above)
Comment (kacf) [reply src/common/http/platform/beast/http.cpp:278]: Not here, unfortunately. `GetContentLength` operates on `optional<size_t>`, whereas this operates on `ExpectedString`. The reason for this difference is that the former comes from data that Beast has interpreted, whereas the latter comes from raw header data.

Comment (lluiscampos) [reply src/common/http/platform/beast/http.cpp:278]: Thank you for clarifying.

File: src/common/http/platform/beast/http.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (kacf) [reply src/common/http/platform/beast/http.cpp]: @lluiscampos: I looked into this a little bit. It's possible to remove the need for using Content-Length in the resumer code, but then we need to change the Range request to not include an end range. The question is whether it's worth it. The resumer is only used for artifacts, where the length typically is known.

Is it needed for the gateway tests?

[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/common/http/platform/beast/http.cpp]: > @lluiscampos: I looked into this a little bit. It's possible to remove the need for using Content-Length in the resumer code, but then we need to change the Range request to not include an end range. The question is whether it's worth it. The resumer is only used for artifacts, where the length typically is known.

Makes sense. Thank you for looking into it.

> Is it needed for the gateway tests?

I don't think wo.

[the remark this answers is not in the document]
Comment (kacf) [reply src/common/http/platform/beast/http.cpp]: Ok, merging this for now then. We can revisit if you run into problems in that area.

**Pull request merged**